design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,542 @@
+better picture of the failure. In addition, it should be associated in a way that allows a
+producing projection for the same language to determine if there is an error info associated
+with the language error and to be able to retrieve it if there is. This will allow the
+producing projection to propagate the error with the previously collected diagnostic information if
+the error ends up being propagated by the component instead of being handled.
+
+It is left to the projections to decide how to achieve this because it will be different for each language.
+But this design note will suggest some approaches to consider based on the error model.
+
+### Exception-based projections
+
+For exception-based projections not affected by object slicing, an approach would be to use
+inheritance. For each of the errors from xlang that the projection maps to, a new type inheriting
+the natural exception type for that error can be created and be used to store the error info. This
+new type would just be an implementation detail of the projection and would not be caught directly
+by components written in that language. Instead components would use the natural exception types
+which these types inherit from to catch exceptions. Projections would use these custom types to
+determine whether it is a translated error and to get the associated error info.
+
+### Error code based projections
+
+For error code based projections, an approach would be to use a thread level storage to store the
+error info. The consuming projection would store the error info there after translating the error,
+while the producing projection would retrieve it from there when the same error propagates to the
+producing side. If the error is handled in the consumer, the consumer should have an API provided
+by the projection to clear the stored error info to ensure proper cleanup of it. The consumer should
+also have an API provided by the projection to get/set the stored error info if the error gets passed
+across threads or if the consumer wants to log it when a failure occurs.
+
+## Getting the information stored in the error info
+
+As mentioned earlier, projections should expose the details stored in the xlang error info as part of
+the function used in the language they project to log errors (i.e. ToString on an exception).
+This will allow consumers to get a full picture of the failure when they log an error using the natural and
+familiar code they would typically use. To achieve this, xlang provides the `IXlangErrorInfo` interface.
+
+```cpp
+interface IXLangErrorInfo
+{
+    void GetError(xlang_result* error);
+    void GetMessage(xlang_string* message);
+    void GetLanguageError(xlang_string* error);
+    void GetExecutionTrace(IXlangErrorExecutionTrace* executionTrace);
+    void GetLanguageIdentifier(xlang_string* identifier);
+    void GetLanguageInformation(IUnknown* languageInformation);
+    void GetLanguageInformationAsString(xlang_string* languageInformation);
```

Comment:
again, this method can go - IStringable from GetLanguageInformation will do

Reply:
Addressed by using the GetObjectInfo string representation category which seems to be what we decided on to address this interface.